src/main/java/net/minecraftforge/resource/ResourcePackLoader.java

```diff
@@ -6,54 +6,97 @@
                 map(IModFileInfo::getFile).map(mf->modResourcePacks.get(mf));
     }
 
-    @Deprecated
-    public static void loadResourcePacks(PackRepository resourcePacks, BiFunction<Map<IModFile, ? extends PathPackResources>, BiConsumer<? super PathPackResources, Pack>, ? extends RepositorySource> packFinder) {
-        loadResourcePacks(resourcePacks, (map) -> packFinder.apply(map, (rp,p) -> {}));
-    }
-
-    public static void loadResourcePacks(PackRepository resourcePacks, Function<Map<IModFile, ? extends PathPackResources>, ? extends RepositorySource> packFinder) {
+    public static void loadResourcePacks(PackRepository resourcePacks, Function<Map<IModFile, ? extends PathPackResources>, ? extends RepositorySource> packFinder, PackType packType) {
         modResourcePacks = ModList.get().getModFiles().stream()
                 .filter(mf->mf.requiredLanguageLoaders().stream().noneMatch(ls->ls.languageName().equals("minecraft")))
-                .map(mf -> Pair.of(mf, createPackForMod(mf)))
+                .map(mf -> Pair.of(mf, createPackForMod(mf, packType)))
                 .collect(Collectors.toMap(p -> p.getFirst().getFile(), Pair::getSecond, (u,v) -> { throw new IllegalStateException(String.format(Locale.ENGLISH, "Duplicate key %s", u)); },  LinkedHashMap::new));
         resourcePacks.addPackFinder(packFinder.apply(modResourcePacks));
     }
 
+    public static RepositorySource buildPackFinder(Map<IModFile, ? extends PathPackResources> modResourcePacks, PackType packType) {
+        return packAcceptor -> packFinder(modResourcePacks, packAcceptor, packType);
+    }
+
+    private static void packFinder(Map<IModFile, ? extends PathPackResources> modResourcePacks, Consumer<Pack> packAcceptor, PackType packType) {
+        var hiddenPacks = new ArrayList<Pack>();
+        for (Map.Entry<IModFile, ? extends PathPackResources> e : modResourcePacks.entrySet())
+        {
+            IModInfo mod = e.getKey().getModInfos().get(0);
+            if (Objects.equals(mod.getModId(), "minecraft")) continue; // skip the minecraft "mod"
+            final String name = "mod:" + mod.getModId();
+            final Pack modPack = Pack.readMetaAndCreate(name, Component.literal(e.getValue().packId()), false, id -> e.getValue(), packType, Pack.Position.BOTTOM, PackSource.DEFAULT);
+            if (modPack == null) {
+                // Vanilla only logs an error, instead of propagating, so handle null and warn that something went wrong
+                ModLoader.get().addWarning(new ModLoadingWarning(mod, ModLoadingStage.ERROR, "fml.modloading.brokenresources", e.getKey()));
+                continue;
+            }
+            LOGGER.debug(Logging.CORE, "Generating PackInfo named {} for mod file {}", name, e.getKey().getFilePath());
+            if ((packType == PackType.CLIENT_RESOURCES && mod.getOwningFile().showAsResourcePack()) || mod.getOwningFile().showAsDataPack()) {
```

Comment:
This doesn't do what you want it to do, it will incorrectly add a resource pack as standalone if the mod says it wants a standalone datapack. You need to either check `packType == PackType.CLIENT_RESOURCES ? mod.getOwningFile().showAsResourcePack() : mod.getOwningFile().showAsDataPack()` or `(packType == PackType.CLIENT_RESOURCES && mod.getOwningFile().showAsResourcePack()) || (packType == PackType.SERVER_DATA && mod.getOwningFile().showAsDataPack())`. The same applies to the `PackResources#isHidden()` implementation in `createPackForMod()`

Reply:
Should be fixed now